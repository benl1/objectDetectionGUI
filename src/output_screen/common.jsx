function drawBoundingBoxes(context, json_response) {
    const num_boxes = json_response['boxes'].length;
    context.strokeStyle = 'red';

    // draw the bounding boxes and their associated scores to the scene image
    for (let i = 0; i < num_boxes; i++) {
        const [top_x, top_y, bot_x, bot_y] = json_response['boxes'][i];
        const confidence = json_response['scores'][i];
        const height = bot_y - top_y;

        context.strokeRect(top_x, top_y, bot_x, bot_y);
        context.strokeText(`${confidence}`, top_x + 5, top_y + height / 2);
    }
}

function resizeHandler(self) {
    console.log('got a resize request');
    const output_width = window.innerWidth * .5;
    const output_height = window.innerHeight * .75;
    console.log(self);
    self.setState({
        output_width: output_width,
        output_height: output_height,
    });
}

export { drawBoundingBoxes, resizeHandler };