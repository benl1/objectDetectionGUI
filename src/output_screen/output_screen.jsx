import { displayImageChoiceScreen, displaySceneSelectionScreen } from '../entrypoint';
import { getImageDataFromURL, produceRGBArray } from '../control/imageops';
import React from 'react';

export default class OutputScreen extends React.Component {
    render() {
        return (
            <div>
                <OutputScreenContainer app={this.props.app}/>
                <div className='button' onClick={() => displayImageChoiceScreen(this.props.app)}>Image choice</div>
                <div className='button' onClick={() => displaySceneSelectionScreen(this.props.app)}>Scene selection</div>
            </div>
        );
    }
}

class OutputScreenContainer extends React.Component {
    constructor(props) {
        super(props);
        this.canvas_ref = React.createRef();
    }

    /**
     * The changes made here are based on this link: 
     */
    componentDidMount() {
        let xhttp = new XMLHttpRequest();
        let target_images = [];
        let response;
    
        // convert image paths to arrays of image data
        this.props.app.images.forEach(img_path => target_images.push(produceRGBArray(getImageDataFromURL(img_path))));
    
        // setup the XHR
        xhttp.open('POST', 'http://127.0.0.1:5000/detect', false);
        xhttp.setRequestHeader('Content-Type', 'application/json');
        xhttp.onreadystatechange = () => {
            if (xhttp.readyState === XMLHttpRequest.DONE && xhttp.status === 200) {
                response = JSON.parse(xhttp.responseText);
            } else if (xhttp.status === 400) {
                displayErrorDialog('server responded with 400.');
            } else {
                displayErrorDialog(`status: ${xhttp.status}, text: ${xhttp.responseText}`);
            }
        };
    
        try {
            let scene = produceRGBArray(getImageDataFromURL(this.props.app.scene_image));
            xhttp.send(JSON.stringify({ scene: scene, targets: target_images }));
        } catch (err) {
            console.log(err);
            //displayErrorDialog('failed to connect to server');
            return; // TODO: for now we can just return prematurely
        }

        const canvas = this.canvas_ref.current;
        const ctx = canvas.getContext('2d');
        const scene_image = new Image();
        
        scene_image.src = this.props.app.scene_image;

        canvas.width = scene_image.naturalWidth;
        canvas.height = scene_image.naturalHeight;

        ctx.drawImage(scene_image, 0, 0);

        ctx.strokeStyle = 'red';
        console.log(response);
        const num_boxes = response['boxes'].length;

        for (let i = 0; i < num_boxes; i++) {
            const [top_x, top_y, bot_x, bot_y] = response['boxes'][i];
            const confidence = response['scores'][i];
            const height = bot_y - top_y;

            ctx.strokeRect(top_x, top_y, bot_x, bot_y);
            ctx.strokeText(`${confidence}`, top_x + 5, top_y + height / 2);
        }
    }

    render() {
        return <canvas ref={this.canvas_ref} />;
    }
}