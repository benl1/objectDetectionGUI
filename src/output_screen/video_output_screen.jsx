import { displayImageChoiceScreen, displaySceneSelectionScreen } from '../entrypoint';
import { drawBoundingBoxes, resizeHandler, BoundingBoxSettings } from './common';
import { WebcamOutput } from '../scene_selection_screen/scene_selection_screen';
import { getImageDataFromURL, produceRGBArray } from '../control/imageops';
import { displayErrorDialog } from '../control/dialogs';
import React from 'react';


const fs = require('fs');

export default function VideoOutputScreen(props) {
    return (
        <div className='flexColumn'>
            <VideoOutputContainer app={props.app}></VideoOutputContainer>
            <div className='button' onClick={() => displayImageChoiceScreen(props.app)}>Image choice</div>
            <div className='button' onClick={() => displaySceneSelectionScreen(props.app)}>Scene selection</div>
        </div>
    );
}

class VideoOutputContainer extends React.Component {
    constructor(props) {
        super(props);
        this.webcam_output_ref = React.createRef();
        this.realtime_video_ref = React.createRef();
        this.box_settings_ref = React.createRef();
        this.firstFrame = true;
        this.inputQueue = [];
        this.outputQueue = [];
        this.recorder = undefined;

        this.chunks = [];

        this.state = {
            output_width: 0,
            output_height: 0,
        }
    }

    componentDidMount() {
        // call the resize handler and register it with the window.

        this.resizeHandler = resizeHandler.bind(this);
        this.resizeHandler();
        window.addEventListener('resize', this.resizeHandler);

        // calculate the RGB arrays for the target images once so we don't have to
        // every frame of video that we go through
        const target_images = [];
        this.props.app.images.forEach(img_path => target_images.push(produceRGBArray(getImageDataFromURL(img_path))));
        this.target_images = target_images;

        this.stream = this.realtime_video_ref.current.getCanvas().captureStream(19)
        var options = {mimeType : 'video/webm;codecs=vp9'}
        this.recorder = new MediaRecorder(this.stream, options)
        this.recorder.ondataavailable = this.handleData
    }

    handleData (e) {
        if (e.data.size > 0){
            this.chunks.push(e.data)
        }
    }
    componentWillUnmount() {
        window.removeEventListener('resize', this.resizeHandler);
    }

    handleFrame(imgData) {
        this.inputQueue.push(imgData); // add new image data to the end of the queue
        if (this.firstFrame) {
            this.firstFrame = false;
            this.processFrame();
            console.log("Started recorder")
            this.recorder.start()
            console.log(this.recorder)

        }
    }

    processFrame() {
        const imgData = this.inputQueue.shift(); // take the image data from the front of the queue
        const rtvComp = this.realtime_video_ref.current;
        const target_images = this.target_images;
        const self = this;

        // draw the image data to a canvas
        const temp_canvas = document.createElement("canvas");
        temp_canvas.width = imgData.width;
        temp_canvas.height = imgData.height;
        const temp_ctx = temp_canvas.getContext('2d');
        temp_ctx.drawImage(imgData, 0, 0, temp_canvas.width, temp_canvas.height);

        // Start the recording if the first frame of the output has been created

        

        // prepare the rest call
        const xhttp = new XMLHttpRequest();
        xhttp.open('POST', 'http://127.0.0.1:5000/detect', true);
        xhttp.setRequestHeader('Content-Type', 'application/json');
        xhttp.onreadystatechange = () => {
            if (xhttp.readyState === XMLHttpRequest.DONE && xhttp.status === 200) {
                const response = JSON.parse(xhttp.responseText);
                const box_settings = {
                    box_count: this.box_settings_ref.current.getNumBoxes(),
                    threshold: this.box_settings_ref.current.getThreshold(),
                }
                drawBoundingBoxes(temp_ctx, response, box_settings);
                this.outputQueue.push(temp_canvas.toDataURL("image/png"))
                rtvComp.showFrame(temp_canvas);

                const callback = () => {
                    if (self.inputQueue.length > 0) {
                        self.processFrame();
                    } else {
                        setTimeout(callback, 100);
                    }
                };

                callback();
            } else if (xhttp.status !== 200) {
                displayErrorDialog('Server error');
            }
        };

        const scene = produceRGBArray(temp_ctx.getImageData(0, 0, temp_canvas.width, temp_canvas.height));
        xhttp.send(JSON.stringify({ scene: scene, targets: target_images }));
    }

    render() {
        const self = this;
        return (
            <div className='flexColumn'>
                <div className='videoOutput'>
                    <WebcamOutput showPausedButton={true} ref={this.webcam_output_ref} parent={this} width={this.state.output_width} height={this.state.output_height}></WebcamOutput>
                    <RealTimeVideo parent={this} ref={this.realtime_video_ref} width={this.state.output_width} height={this.state.output_height}></RealTimeVideo>
                </div>
                <div className='button' onClick={(e) => {
                    e.target.innerHTML = self.webcam_output_ref.current.pause() ? "Resume video" : "Pause video";
                }}>Pause video</div>
                <BoundingBoxSettings ref={this.box_settings_ref}></BoundingBoxSettings>
                <div id = "download_video" className="button" onClick = { () => {
                        self.recorder.stop()
                        var blob = new Blob(self.chunks, {type: "video/webm"})  
                        var urlObj = URL.createObjectURL(blob)
                        video_button = document.getElementById("download_video")
                        video_button.href = urlObj
                        video_button.download = "test.webm"
                        video_button.click()
                        window.URL.revokeObjectURL(urlObj)

                    }
                }></div>
            </div>
        );
    }
}

class RealTimeVideo extends React.Component {
    constructor(props) {
        super(props);
        this.canvas_ref = React.createRef();
    }
    getCanvas() {
        return this.canvas_ref.current;
    }
    componentDidMount() {
        const canvas = this.canvas_ref.current;
        canvas.width = this.props.width;
        canvas.height = this.props.height;

        const ctx = canvas.getContext('2d');
        const load = new Image();
        load.onload = () => ctx.drawImage(load, 0, 0, canvas.width, canvas.height);
        load.src = "./assets/loading.jpg";
    }

    showFrame(imgData) {
        const canvas = this.canvas_ref.current;
        canvas.width = this.props.width;
        canvas.height = this.props.height;
        const ctx = canvas.getContext('2d');
        ctx.drawImage(imgData, 0, 0, canvas.width, canvas.height);
    }

    render() {
        return <div><canvas ref={this.canvas_ref}></canvas></div>;
    }
}