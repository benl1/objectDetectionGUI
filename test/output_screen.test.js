// output_screen.test.js
import React from 'react';
import { shallow, mount } from 'enzyme';
import OutputScreen from '../src/output_screen/output_screen';
import { OutputScreenContainer } from '../src/output_screen/output_screen';
import renderer from 'react-test-renderer';
import App from '../src/control/app';

describe('Output screens', () => {

    it('test that outputscreen renders correctly', ()=>{
        let component = shallow(<OutputScreen app={new App()}/>);
        expect(component.find(OutputScreenContainer)).toHaveLength(1);
        expect(component.find('.button')).toHaveLength(2);
        expect(component).toMatchSnapshot();
    })

    it('test that outputscreencontainer renders correctly', ()=>{
        let component = mount(<OutputScreenContainer app={new App()}/>)
        expect(component.find('.videoOutput')).toHaveLength(1);
        expect(component).toMatchSnapshot();
    })

});