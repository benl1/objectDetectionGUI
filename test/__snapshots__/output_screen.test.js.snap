// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Output screens test that outputscreen renders correctly 1`] = `
<div
  className="flexColumn"
>
  <OutputScreenContainer
    app={
      App {
        "image_key": 0,
        "images": Array [],
      }
    }
  />
  <div
    className="button"
    onClick={[Function]}
  >
    Image choice
  </div>
  <div
    className="button"
    onClick={[Function]}
  >
    Scene selection
  </div>
</div>
`;

exports[`Output screens test that outputscreencontainer renders correctly 1`] = `
<OutputScreenContainer
  app={
    App {
      "image_key": 0,
      "images": Array [],
    }
  }
>
  <div
    className="flexColumn"
  >
    <div
      className="videoOutput"
    >
      <canvas
        className="dontGrow"
      />
      <canvas
        className="dontGrow"
      />
    </div>
  </div>
</OutputScreenContainer>
`;
